lAyeR/lAyeRTests/Vector3DTests.swift

```diff
@@ -33,12 +38,63 @@ class Vector3DTests: XCTestCase {
                        "The opposite-direction vector of \(v) is incorrect!")
     }
     
+    func testNormalDotProduct() {
+        let v1 = Vector3D(x: 1, y: 2, z: 3)
+        let v2 = Vector3D(x: 3, y: 2, z: 1)
+        XCTAssertEqual(v1 * v2, 10,
+                       "The dot product is incorrect!")
+
+    }
+    
     func testDotProductWithZeroVector() {
         let zero = Vector3D(x: 0, y: 0, z: 0)
         let v = Vector3D(x: 1, y: 3, z: 9)
         XCTAssertEqual(zero * v, 0,
                        "The dot product with zero vector is incorrect!")
     }
+    
+    func testDotProductWithPerpendicularVector() {
+        let x = Vector3D(x: 1, y: 0, z: 0)
+        let y = Vector3D(x: 0, y: 1, z: 0)
+        let z = Vector3D(x: 0, y: 0, z: 1)
+        XCTAssertEqual(x * y, 0,
+                       "The dot product between 2 perpendicular vectors is not 0!")
+
+        XCTAssertEqual(y * z, 0,
+                       "The dot product between 2 perpendicular vectors is not 0!")
+
+        XCTAssertEqual(x * z, 0,
+                       "The dot product between 2 perpendicular vectors is not 0!")
+
+    }
+    
+    func testNormalProjectionLength() {
+        let v1 = Vector3D(x: 1, y: 1, z: 1)
+        let v2 = Vector3D(x: 1, y: 1, z: 0)
+        print(v1.projectionLength(on: v2))
+        XCTAssertEqualWithAccuracy(v1.projectionLength(on: v2), sqrt(2), accuracy: 1e-8)
+    }
+    
+    func testProjectionLengthOnZeroVector() {
+        let zero = Vector3D(x: 0, y: 0, z: 0)
+        let v = Vector3D(x: 1, y: 3, z: 9)
+        XCTAssertEqual(v.projectionLength(on: zero), 0)
+    }
+    
+    func testProjectionOfZero() {
+        let zero = Vector3D(x: 0, y: 0, z: 0)
+        let v = Vector3D(x: 1, y: 3, z: 9)
+        XCTAssertEqual(zero.projectionLength(on: v), 0)
+    }
+    
+    func testProjectionLengthOnPerpendicularVector() {
```

Comment:
see the most recent commit:
the first is for testing dot product
the second is now changed to test projection